Mathlib/Tactic/LiftLets.lean

```diff
@@ -21,6 +21,10 @@ This list is used during the computation to merge let bindings.
 -/
 private partial def Lean.Expr.liftLetsAux (e : Expr) (fvars : Array Expr)
     (f : Array Expr → Expr → MetaM Expr) : MetaM Expr := do
+  if (e.find? Expr.isLet).isNone then
+    return ← f fvars e
```

Comment:
Is this meaningfully better at avoiding deep recursion vs just letting the match fall through?

Reply:
What do you mean "*just* letting the match fall through"? This is skipping looking at `e` if it doesn't contain a `let` anywhere, which means avoiding for example processing `.app`s or `.forallE`s in such expressions.

Reply:
I guess my claim is that `find?` is implemented by an analogous `match`, and the processing overhead on the negative path in your code doesn't look very expensive; but perhaps I'm missing something.

Were you able to check which of the two `if`s is needed to make the golfed proof possible? Do either of them work in isolation, or is only both enough to give the needed speed boost?

Reply:
It's not about the depth of expression -- `lift_lets` is implemented using continuation-passing style, so it's the total size of the expression visited that's contributes to recursion depth. Plus, there's a good amount of processing that'd be nice to skip if a subexpression doesn't have any let bindings.

Reply:
I assume there's no easy way to rewrite the continuation-passing approach to fit within the monad? (or if there is, it has other performance issues)

Reply:
There's probably a way to write it differently, but the current way seemed right since each piece of an expression needs to be able to create let bindings that are visible to the next subexpressions being processed. Continuation-passing style is prevalent in meta code that manipulates local contexts.